packages/core/src/browser/tree/tree-selection.ts

```diff
@@ -67,45 +100,109 @@ export namespace ISelectableTreeNode {
 
     dispose() {
         this.onSelectionChangedEmitter.dispose();
     }
 
-    get selectedNode(): ISelectableTreeNode | undefined {
-        return this._selectedNode;
+    get selectedNodes(): TreeSelection {
+        return this._selectedNodes.slice();
     }
 
-    get onSelectionChanged(): Event<ISelectableTreeNode | undefined> {
+    get onSelectionChanged(): Event<TreeSelection> {
         return this.onSelectionChangedEmitter.event;
     }
 
     protected fireSelectionChanged(): void {
-        this.onSelectionChangedEmitter.fire(this._selectedNode);
+        this.onSelectionChangedEmitter.fire(this._selectedNodes.slice());
     }
 
-    selectNode(raw: ISelectableTreeNode | undefined): void {
-        const node = this.tree.validateNode(raw);
+    selectNode(raw: ISelectableTreeNode, options?: ITreeSelectionService.Options): void {
+        const node = this.validateNode(raw);
         if (ISelectableTreeNode.is(node)) {
-            this.doSelectNode(node);
-        } else {
-            this.doSelectNode(undefined);
+            this.doSelectNode(node, options);
         }
     }
 
-    protected doSelectNode(node: ISelectableTreeNode | undefined): void {
-        if (this._selectedNode) {
-            this._selectedNode.selected = false;
+    protected doSelectNode(node: ISelectableTreeNode, options?: ITreeSelectionService.Options): void {
+        // Shortcut. Nothing is selected yet. Select the node and we are done.
+        if (this._selectedNodes.length === 0) {
+            node.selected = true;
+            this._selectedNodes.unshift(node);
+            this.fireSelectionChanged();
+            return;
         }
-        this._selectedNode = node;
-        if (node) {
+
+        const multi = options && options.multi;
+        const index = this._selectedNodes.indexOf(node);
+        let changed = false;
+
+        if (multi) {
+            // The node is not yet selected;
+            if (index === -1) {
+                node.selected = true;
+                this._selectedNodes.unshift(node);
+                changed = true;
+            } else {
+                // length === 0 => We already covered this case. Nothing to do.
+                // length === 1 => The node we wan to select is already selected. Nothing to do.
+                // length !== 1 => We need to unlink the selected node and put it into the start position.
+                if (this._selectedNodes.length !== 1) {
+                    node.selected = true;
+                    this._selectedNodes.splice(index, 1);
+                    this._selectedNodes.unshift(node);
+                    changed = true;
+                }
+            }
+        } else if (this._selectedNodes.length !== 1 || index === -1) {
+            this.clearSelections();
             node.selected = true;
+            this._selectedNodes.unshift(node);
+            changed = true;
         }
-        this.fireSelectionChanged();
```

Comment:
it used to be called always. I think there was a reason for it.

Reply:
If the selection does not change, what is the benefit?

Reply:
don't know, ok, let's try to keep like that